torch/cuda/amp/grad_scaler.py

```diff
@@ -336,6 +336,8 @@ def step(self, optimizer, *args, **kwargs):
             # and `found_inf` to the passed optimizer so that the optimizer can utilize those
             # to skip the parameter updates or unscale gradients before updating parameters in
             # the fused kernel, e.g. `FusedAdamMathFunctor`.
+            if optimizer_state["stage"] is OptState.READY:
+                self._check_inf_per_device(optimizer)
             kwargs_ = kwargs
             has_grad_scaler_kwarg = "grad_scaler" in inspect.signature(optimizer.step).parameters
             if has_grad_scaler_kwarg:
```

Comment:
Just confirming--is it right to say that this branch does not need updates because other optimizers (like the apex ones) already handle the check_inf/nan order for the scaled grads only?

Reply:
updates of what?

if `optimizer._step_supports_amp_scaling` is `True`, then the optimizer takes care of both unscaling gradients and parameter updates

Reply:
But it shouldn't re-check again if the unscaling already happened right?

Reply:
this if statement is to avoid the unwanted re-check of gradients. `GradScaler.unscale_(optimizer)` updates `OptState` to `OptState.UNSCALED`: https://github.com/pytorch/pytorch/blob/d8d479c854c622a6ef21190b7e62c9a76f0ea2a7/torch/cuda/amp/grad_scaler.py#L285

Reply:
As long as the future optimizers that would fall under the has_grad_scaler_kwarg already reuse the calculated found_inf dictionary instead of recomputing, this would be true. 

I was asking for a confirmation because I am not familiar with the implementations of those optimizers, though it looks like we don't define a contract on our side.

Reply:
ah, would you mean `_check_inf_per_device` should not be called if `has_grad_scaler_kwarg` because we have no control in that path?

Reply:
It depends on what our contract is, though it looks like it hasn’t been defined previously. I did a GH search and saw that apex fused adam would check the infs in spite of the unscale_ api (see https://github.com/NVIDIA/apex/blob/817e8187fc3960985af89bd09094f08082307080/apex/optimizers/fused_adam.py#L204). 

So I guess it does not really matter right now where the check inf goes, though it looks more correct to be where you’ve put it now (under the specific branch). I’m not sure if we want to establish the contract that the grad scaler will set the state to UNSCALED vs READY and have optimizers condition on those states. Since that path is already in deprecation stages, I do not think it is high priority. Feel free to land this!